File: android_app/app/src/main/java/com/oceansentinels/app/presentation/viewmodel/AuthViewModel.kt
package com.oceansentinels.app.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.oceansentinels.app.domain.model.*
import com.oceansentinels.app.domain.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * ViewModel for authentication-related operations
 */
@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    // User session state
    val userSession: StateFlow<UserSession> = authRepository.getUserSession()
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = UserSession.Loading
        )

    // Login state
    private val _loginState = MutableStateFlow<LoginState>(LoginState.Idle)
    val loginState: StateFlow<LoginState> = _loginState.asStateFlow()

    // Register state
    private val _registerState = MutableStateFlow<RegisterState>(RegisterState.Idle)
    val registerState: StateFlow<RegisterState> = _registerState.asStateFlow()

    // Current user
    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser.asStateFlow()

    init {
        observeUserSession()
    }

    private fun observeUserSession() {
        viewModelScope.launch {
            userSession.collect { session ->
                when (session) {
                    is UserSession.Authenticated -> {
                        _currentUser.value = session.user
                    }
                    else -> {
                        _currentUser.value = null
                    }
                }
            }
        }
    }

    /**
     * Login with username and password
     */
    fun login(username: String, password: String) {
        viewModelScope.launch {
            _loginState.value = LoginState.Loading
            
            val result = authRepository.login(
                LoginCredentials(username = username, password = password)
            )
            
            result.fold(
                onSuccess = { authToken ->
                    Timber.d("Login successful for user: ${authToken.user.username}")
                    _loginState.value = LoginState.Success(authToken.user)
                },
                onFailure = { error ->
                    Timber.e(error, "Login failed")
                    _loginState.value = LoginState.Error(error.message ?: "Login failed")
                }
            )
        }
    }

    /**
     * Demo login for testing
     */
    fun demoLogin(role: UserRole) {
        viewModelScope.launch {
            _loginState.value = LoginState.Loading
            
            val demoUser = when (role) {
                UserRole.PUBLIC -> LoginCredentials("sihcitizen@vi.com", "SIH@2025")
                UserRole.ADMIN -> LoginCredentials("OceanAdmin1", "admin")
                UserRole.RESCUE_TEAM -> LoginCredentials("sihrescue@vi.com", "Ocean@123")
                UserRole.AUTHORITY -> LoginCredentials("sihauthority@vi.com", "Ocean@123")
            }
            
            login(demoUser.username, demoUser.password)
        }
    }

    /**
     * Register a new user
     */
    fun register(
        username: String,
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        phone: String? = null,
        location: String? = null
    ) {
        viewModelScope.launch {
            _registerState.value = RegisterState.Loading
            
            val result = authRepository.register(
                RegistrationData(
                    username = username,
                    email = email,
                    password = password,
                    firstName = firstName,
                    lastName = lastName,
                    phone = phone,
                    location = location,
                    role = UserRole.PUBLIC
                )
            )
            
            result.fold(
                onSuccess = { user ->
                    Timber.d("Registration successful for user: ${user.username}")
                    _registerState.value = RegisterState.Success(user)
                },
                onFailure = { error ->
                    Timber.e(error, "Registration failed")
                    _registerState.value = RegisterState.Error(error.message ?: "Registration failed")
                }
            )
        }
    }

    /**
     * Logout the current user
     */
    fun logout() {
        viewModelScope.launch {
            authRepository.logout()
            _currentUser.value = null
            _loginState.value = LoginState.Idle
        }
    }

    /**
     * Reset login state
     */
    fun resetLoginState() {
        _loginState.value = LoginState.Idle
    }

    /**
     * Reset register state
     */
    fun resetRegisterState() {
        _registerState.value = RegisterState.Idle
    }

    /**
     * Check if user is authenticated
     */
    suspend fun isAuthenticated(): Boolean {
        return authRepository.isAuthenticated()
    }
}

/**
 * Login state sealed class
 */
sealed class LoginState {
    data object Idle : LoginState()
    data object Loading : LoginState()
    data class Success(val user: User) : LoginState()
    data class Error(val message: String) : LoginState()
}

/**
 * Register state sealed class
 */
sealed class RegisterState {
    data object Idle : RegisterState()
    data object Loading : RegisterState()
    data class Success(val user: User) : RegisterState()
    data class Error(val message: String) : RegisterState()
}

File: IOS/OceanSentinels/Presentation/ViewModels/AuthViewModel.swift
import Foundation
import Observation

// MARK: - AuthViewModel

/// ViewModel for authentication — login, registration, logout, session management.
@Observable
@MainActor
final class AuthViewModel {

    // MARK: - State

    var userSession: UserSession = .loading
    var loginState: LoginState = .idle
    var registerState: RegisterState = .idle
    var currentUser: User?

    // MARK: - Dependencies

    private let authRepository: any AuthRepository

    // MARK: - Init

    init(authRepository: any AuthRepository) {
        self.authRepository = authRepository
        self.currentUser = nil

        Task { await observeUserSession() }
    }

    // MARK: - Session Observation

    private func observeUserSession() async {
        let session = authRepository.getUserSession()
        self.userSession = session
        if case .authenticated(let user, _) = session {
            self.currentUser = user
        } else {
            self.currentUser = nil
        }
    }

    // MARK: - Login

    /// Login with username and password.
    func login(username: String, password: String) {
        Task {
            loginState = .loading

            do {
                let authToken = try await authRepository.login(
                    credentials: LoginCredentials(username: username, password: password)
                )
                AppLogger.auth.info("Login successful for user: \(authToken.user.username)")
                loginState = .success(authToken.user)
                currentUser = authToken.user
                userSession = .authenticated(user: authToken.user, token: authToken.accessToken)
            } catch {
                AppLogger.auth.error("Login failed: \(error.localizedDescription)")
                loginState = .error(error.localizedDescription)
            }
        }
    }

    /// Demo login for testing.
    func demoLogin(role: UserRole) {
        let credentials: (String, String) = switch role {
        case .public:
            ("sihcitizen@vi.com", "SIH@2025")
        case .admin:
            ("OceanAdmin1", "admin")
        case .rescueTeam:
            ("sihrescue@vi.com", "Ocean@123")
        case .authority:
            ("sihauthority@vi.com", "Ocean@123")
        }
        login(username: credentials.0, password: credentials.1)
    }

    // MARK: - Registration

    /// Register a new user.
    func register(
        username: String,
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        phone: String? = nil,
        location: String? = nil
    ) {
        Task {
            registerState = .loading

            do {
                let user = try await authRepository.register(
                    data: RegistrationData(
                        username: username,
                        email: email,
                        password: password,
                        firstName: firstName,
                        lastName: lastName,
                        phone: phone,
                        location: location,
                        role: .public
                    )
                )
                AppLogger.auth.info("Registration successful for user: \(user.username)")
                registerState = .success(user)
            } catch {
                AppLogger.auth.error("Registration failed: \(error.localizedDescription)")
                registerState = .error(error.localizedDescription)
            }
        }
    }

    // MARK: - Logout

    func logout() {
        Task {
            await authRepository.logout()
            currentUser = nil
            userSession = .unauthenticated
            loginState = .idle
        }
    }

    // MARK: - State Resets

    func resetLoginState() {
        loginState = .idle
    }

    func resetRegisterState() {
        registerState = .idle
    }

    // MARK: - Queries

    func isAuthenticated() -> Bool {
        return authRepository.isAuthenticated()
    }
}

// MARK: - Login State

enum LoginState: Equatable {
    case idle
    case loading
    case success(User)
    case error(String)

    static func == (lhs: LoginState, rhs: LoginState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle), (.loading, .loading): return true
        case (.success(let a), .success(let b)): return a.id == b.id
        case (.error(let a), .error(let b)): return a == b
        default: return false
        }
    }
}

// MARK: - Register State

enum RegisterState: Equatable {
    case idle
    case loading
    case success(User)
    case error(String)

    static func == (lhs: RegisterState, rhs: RegisterState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle), (.loading, .loading): return true
        case (.success(let a), .success(let b)): return a.id == b.id
        case (.error(let a), .error(let b)): return a == b
        default: return false
        }
    }
}
